src/cargo/core/resolver/mod.rs

```diff
@@ -774,18 +757,50 @@ impl RemainingCandidates {
             // semver-compatible versions of a crate. For example we can't
             // simultaneously activate `foo 1.0.2` and `foo 1.2.0`. We can,
             // however, activate `1.0.2` and `2.0.0`.
-            //
-            // Here we throw out our candidate if it's *compatible*, yet not
-            // equal, to all previously activated versions.
             if let Some((a, _)) = cx.activations.get(&b_id.as_activations_key()) {
-                if *a != b {
-                    conflicting_prev_active
+                // If this candidate is already activated, then we want to put
+                // it in our prioritized list to try first.
+                if a == b {
+                    activated_candidates.push(b.clone());
+                    continue;
+                }
+                // Here we throw out our candidate if it's *compatible*, yet not
+                // equal, to all previously activated versions.
+                else {
+                    conflicting_activations
                         .entry(a.package_id())
                         .or_insert(ConflictReason::Semver);
                     continue;
                 }
+            } else {
+                non_activated_candidates.push(b.clone());
             }
+        }
 
+        // Combine the prioritized and non-prioritized candidates into one list
+        // such that the prioritized candidates are tried first.
+        activated_candidates.append(&mut non_activated_candidates);
+
+        (
+            RemainingCandidates {
+                prioritized_candidates: RcVecIter::new(Rc::new(activated_candidates)),
+                has_another: None,
+            },
+            conflicting_activations,
+        )
+    }
+
+    /// Attempts to find another candidate to check from this list.
+    ///
+    /// This method will attempt to move this iterator forward, returning a
+    /// candidate that's possible to activate.
+    ///
+    /// If successful a `(Candidate, bool)` pair will be returned. The
+    /// `Candidate` is the candidate to attempt to activate, and the `bool` is
+    /// an indicator of whether there are remaining candidates to try of if
+    /// we've reached the end of iteration.
+    fn next(&mut self) -> Option<(Summary, bool)> {
```

Comment:
if next does not take additional arguments, then it should probably be a proper iterator. ( Or document why it can't implement that trait)